homeassistant/components/snooz/__init__.py

```diff
@@ -60,3 +61,103 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
             hass.data.pop(DOMAIN)
 
     return unload_ok
+
+
+async def _async_load_entry_data(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> SnoozConfigurationData:
+    if _entry_missing_device_config(entry):
+        adv_data = await _async_update_entry_config(hass, entry)
+    else:
+        adv_data = _load_entry_config(entry)
+
+    address: str = entry.data[CONF_ADDRESS]
+    if not (ble_device := async_ble_device_from_address(hass, address)):
+        raise ConfigEntryNotReady(
+            f"Could not find Snooz with address {address}. Try power cycling the device"
+        )
+
+    return SnoozConfigurationData(
+        ble_device,
+        adv_data,
+        device=SnoozDevice(ble_device, adv_data),
+        title=entry.title,
+    )
+
+
+def _entry_missing_device_config(entry: ConfigEntry) -> bool:
+    return CONF_MODEL not in entry.data or CONF_FIRMWARE_VERSION not in entry.data
+
+
+def _load_entry_config(entry: ConfigEntry) -> SnoozAdvertisementData:
+    return SnoozAdvertisementData(
+        model=SnoozDeviceModel(entry.data[CONF_MODEL]),
+        firmware_version=SnoozFirmwareVersion(entry.data[CONF_FIRMWARE_VERSION]),
+        password=entry.data[CONF_TOKEN],
+    )
+
+
+async def _async_update_entry_config(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> SnoozAdvertisementData:
+    """Find a supported advertisement and store data in config entry."""
+
+    address = entry.data[CONF_ADDRESS]
+    LOGGER.debug(f"Loading device information from advertisement for address {address}")
+
+    adv_data = await hass.async_create_task(
+        _async_wait_for_advertisement(hass, address)
+    )
+
+    if adv_data is None:
+        raise ConfigEntryNotReady(
+            f"Could not find supported advertisement for address {address}."
+            " Try power cycling the device"
+        )
+
+    hass.config_entries.async_update_entry(
+        entry,
+        data={
+            **entry.data,
+            CONF_MODEL: adv_data.model,
+            CONF_FIRMWARE_VERSION: adv_data.firmware_version,
+        },
+    )
+
+    # use token from config entry since we don't require
+    # pairing mode to load the data from an advertisement
+    adv_data.password = entry.data[CONF_TOKEN]
+
+    return adv_data
+
+
+async def _async_wait_for_advertisement(
+    hass: HomeAssistant, address: str
+) -> SnoozAdvertisementData | None:
+    """Process advertisements for an address until a supported advertisement is found."""
+
+    def is_supported_advertisement(
+        service_info: BluetoothServiceInfo,
+    ) -> bool:
+        if parse_snooz_advertisement(service_info) is None:
+            LOGGER.warning(
+                f"Skipping unsupported advertisement with name: {service_info.name}"
+            )
+            return False
+
+        return True
```

Comment:
It looks like this doesn't need to be an inner function and can be moved outside this function.